cosmos/runtime/txpool/mempool.go

```diff
@@ -104,10 +113,30 @@ func (m *Mempool) Insert(ctx context.Context, sdkTx sdk.Tx) error {
 		return errors.New("only one message is supported")
 	}
 
-	if wet, ok := utils.GetAs[*types.WrappedEthereumTransaction](msgs[0]); !ok {
+	wet, ok := utils.GetAs[*types.WrappedEthereumTransaction](msgs[0])
+	if !ok {
 		// We have to return nil for non-ethereum transactions as to not fail check-tx.
 		return nil
-	} else if errs := m.txpool.Add(
+	}
+
+	// If we already have the transaction in the txpool, we can return early.
+	ethTx := wet.Unwrap()
+	if ethTx == nil {
+		return errors.New("wraped tx is nil")
+	}
+
+	// If the tx is a local, or has been gossiped again for some reason. We ignore it.
+	ethTxHash := ethTx.Hash()
+	if m.txpool.Has(ethTxHash) {
+		return nil
+	}
```

Comment:
Has takes the read lock on txpool, so will block miner from building block (both `Pending` and `Locals` do a full RW lock in worker building block). In stage 1 its fine bc both this call and `txpool.Add` lock on mempool. but in stage2 we fully optimize by doing the `pauseInserts`, which means removing the `.Has` check